### ISSUE TITLE
Training iterations and training loss

### ISSUE THREAD

**Issue by DataCorrupted** [google/ml-compiler-opt#451]
Current training script setup (see code below) doesn't seem to care about loss. 
The logic is more "Train for X iterations then stop"

https://github.com/google/ml-compiler-opt/blob/213671606d4a4af0e52c04ff1cd095dd103321ee/compiler_opt/rl/trainer.py#L201-L232

This is somewhat problematic for us:

1. It takes really long to train, maybe the model is pretty good after 1/2 X iterations, yet we still have to wait.
2. When the model is trained, we are not so confident how good it is.  It could be that the model is still not trained enough (e.g. it needs 2X iterations to learn everything), or there could be something fundamentally wrong about the data.

One solution I saw is using a moving average of the loss, stop once it is lower than a threshold (or it exhausted X iterations). |
I want to start a discussion here whether we should do that in the script?

**Comment by alekh** [comment]
Since this is RL training, we do not know apriori how much reward is "good enough". Unlike in classification of regression problems, where we can say what is an acceptable value of loss, that is less clear here. So having a loss-based termination might not be great, since a generic threshold is not obvious here. That said, if you have a clear threshold for your use-case, please feel free to modify in the suggested manner.

**Comment by DataCorrupted** [comment]
That argument doesn't make current setting "run for X iteratiorations" easier, with our prior it is also unclear what "X" is appropriate. I guess the root question here is, how do we define a "good", "trained" model? 
You are right that in classification or regression problems we can say loss close to 0 is good, but not the case here - I've seen negative loss during my training. What does loss represent here?

Engineering wise, maybe we can add an option "loss_threashold" such that when provided, the training is stopped after the running average loss is lower than that.

**Comment by boomanaiden154** [comment]
If you're looking at the reward graphs, the reward is the percent reduction in size/percent reduction in cycles/time for inlining and regalloc respectively. A negative reward just means that what was being measured regressed and is pretty common in the initial stages of training, especially without warmstarting.

The normal way to run these scripts if time is sensitive is to start them, and then check tensorboard until you're happy with the models convergence, and then control+c the training process. It doesn't really make sense from an ML perspective to have some threshold in the gin files to stop training afterwards because you don't know beforehand how long the model is going to take to converge/what the level is. In a continuous training scenario (ie redeploying new models every x weeks/months), you would just set the total number of steps to a reasonable amount above what it takes to converge and train for that amount of time.

**Comment by mtrofin** [comment]
I think adding a threshold would be fine, basically why not have both ways (i.e. if folks want threshold -> fine; if not -> also fine). If it turns out that the threshold isn't really useful, doesn't seem like something that'd be hard to live with if not removed.

@DataCorrupted would that work?

Separately, about iterations - it should also be possible to set the iterations to something economical (small-ish, like few 100K); then re-start training using the model that came out of the first set of iterations as warmstart.

I'm wondering - @alekh @boomanaiden154 - would it be a bad idea to optionally automatically monitor the rate of improvement in the reward and say something like "if it's not improving more than X after some nr of iterations -> stop"?

**Comment by DataCorrupted** [comment]
That sounds good to me.

Combining your first two thoughts, maybe we could do "default to X iterations unless user specified an error threshold (via one flag) or # of iterations (via another flag)"

**Comment by alekh** [comment]
I haven't monitored your training runs carefully enough with PPO to see how
continuously and monotonically the loss improves. You could put in a
heuristic like this, but I'm generally concerned about more hyper
parameters which get added in doing so.

Alekh

**Comment by mtrofin** [comment]
Ack - I was thinking of the heuristic more long-term. Short term, though, @DataCorrupted if you have a patch for the `loss_threshold`, happy to review. Once you got some mileage out of it, it'd be great if you could comment on its usefulness - maybe your setup is different from what we're doing (which is basically what @boomanaiden154 said) and it'd be good to learn from your scenario.

**Comment by DataCorrupted** [comment]
Closing this as `loss_threshold` is not working for us as I imagined would be.

Initially we thought it was because loss data was not used correctly that the model can be trained with less/more time. After some manual adjust to `num_policy_iterations` and `num_iterations` we find that the model behaved similarly.

(BTW: not an ML expert, what is the difference between these two?)

**Comment by alekh** [comment]
> (BTW: not an ML expert, what is the difference between these two?)

num_iteations is the number of gradient updates which are done using the data collected at a fixed policy, and num_policy_iterations is the total number of times we update the data collection policy. So the total number of training steps is num_policy_iterations * num_iterations.
